### DISCUSSION TITLE
How can I upload with multiple files to another server to handle?

### DISCUSSION THREAD

**Discussion by mvn-vyvo-dn** [langgenius/dify#10409] [Help]
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

Can not upload with File List in Http request(form-data), right? How can I use Http request to upload multiple files with input is a File List?
I want to upload multiple files then call api from other service to handle that multiple files. How can I do it with workflow? (Now I use Http request(form-data) but it not working)

**Comment by ShineiArakawa** [comment]
I encountered the same issue with `v0.11.2`.
After debugging, I identified that the issue seems to be in the section under *api/core*.
When I applied the following two changes, it worked for me:

### 1. Modification at [`VariablePool`](https://github.com/langgenius/dify/blob/7f1fdb774c4fa304b67a3d914037edecd31c93d6/api/core/workflow/entities/variable_pool.py#L22) class in `core/workflow/entities/variable_pool.py`

Add a new member function named `get_files` to the `VariablePool` class.

```python
# from core.variables import ArrayFileVariable

def get_files(self, selector: Sequence[str], /) -> ArrayFileVariable | None:
        segment = self.get(selector)
        if isinstance(segment, ArrayFileVariable):
            return segment
        return None
```

### 2. Modification at [`Executor._init_body`](https://github.com/langgenius/dify/blob/7f1fdb774c4fa304b67a3d914037edecd31c93d6/api/core/workflow/nodes/http_request/executor.py#L100) function in `core/workflow/nodes/http_request/executor.py`

- Call `get_files` instead of `get_file`
- Add an unpack operation

***From***
```python
                case "form-data":
                    form_data = {
                        self.variable_pool.convert_template(item.key).text: self.variable_pool.convert_template(
                            item.value
                        ).text
                        for item in filter(lambda item: item.type == "text", data)
                    }
                    file_selectors = {
                        self.variable_pool.convert_template(item.key).text: item.file
                        for item in filter(lambda item: item.type == "file", data)
                    }
                    files = {k: self.variable_pool.get_file(selector) for k, selector in file_selectors.items()}
                    files = {k: v for k, v in files.items() if v is not None}
                    files = {k: variable.value for k, variable in files.items()}
                    files = {
                        k: (v.filename, file_manager.download(v), v.mime_type or "application/octet-stream")
                        for k, v in files.items()
                        if v.related_id is not None
                    }
                    self.data = form_data
                    self.files = files or None
```

***To***
```python
                case "form-data":
                    form_data = {
                        self.variable_pool.convert_template(item.key).text: self.variable_pool.convert_template(
                            item.value
                        ).text
                        for item in filter(lambda item: item.type == "text", data)
                    }
                    file_selectors = {
                        self.variable_pool.convert_template(item.key).text: item.file
                        for item in filter(lambda item: item.type == "file", data)
                    }
                    files = {k: self.variable_pool.get_files(selector) for k, selector in file_selectors.items()} # Call `get_files` instead of `get_file`
                    files = {k: v for k, v in files.items() if v is not None}
                    files = {k: variable.value for k, variable in files.items()}
                    files = {k: vv for k, v in files.items() for vv in v} # Add an unpack operation
                    files = {
                        k: (v.filename, file_manager.download(v), v.mime_type or "application/octet-stream")
                        for k, v in files.items()
                        if v.related_id is not None
                    }
                    self.data = form_data
                    self.files = files or None
```

However, please note that this is just a temporary workaround. I believe this issue might need to be formally reported as an Issue to gather more insights and work towards a more permanent solution.
